### PR TITLE
TS-4212: Add option to track memory allocation with OpenSSL

### PR DESCRIPTION
**Pull request by bryancall**

### CODE REVIEW

File: lib/ts/ink_memory.cc [hunk 1]
```diff
@@ -209,6 +210,46 @@ ats_mlock(caddr_t addr, size_t len)
...
 
+void *
+ats_track_malloc(size_t size, uint64_t *stat)
+{
+  // pad with 8 bytes to add allocation size
+  void *ptr = malloc(size + 8);
+  memcpy(ptr, &size, 8);
+  ink_atomic_increment(stat, size);
+  return (void *)((uint64_t)ptr + 8);
+}
```
Comment (PSUdaemon) [inline lib/ts/ink_memory.cc]: I know we only support 64 bit, but shouldn't we still use `sizeof()` and `uintptr_t`?

Comment (PSUdaemon) [inline lib/ts/ink_memory.cc]: Also, should we use `ats_malloc()` and friends in all these?

**Comment by bryancall** [comment]
I am planning on having and option for {{ats_malloc}} to call the track functions in the future.  Still thinking about how to do it cleanly.

File: lib/ts/ink_memory.cc (hunk 1 above)
Comment (bryancall) [inline lib/ts/ink_memory.cc]: I am planning on having and option for `ats_malloc` to call the track functions in the future. Still thinking about how to do it cleanly.

**Comment by bryancall** [comment]
I updated it with Phil's suggestion.

I can figure out later what to do about calling ats_malloc like function within these tracking functions when I add support for ats_malloc to use the tracking functions.  If that make any sense...

**Comment by PSUdaemon** [comment]
:+1: This looks good to me.

**Comment by bgaff** [comment]
What about the equivalents to posix_memalign?

**Comment by jpeach** [comment]
Are you sure there isn't a better way to do this?

- it is fairly ugly
- it duplicates the allocation size tracking the allocator is already doing
- it gets the alignment wrong in some cases (ie. you won't be 16 byte aligned any more).

[glibc](http://man7.org/linux/man-pages/man3/malloc_usable_size.3.html) and [jemalloc](https://www.freebsd.org/cgi/man.cgi?query=malloc_usable_size) both provide ``malloc_usable_size``, which seems like it will tell you what you need to know here.

**Comment by PSUdaemon** [comment]
@jpeach,

I can't speak to the relative ugliness of the code. It's fairly C-ish like the rest of our memory related code.

I don't think the current implementation of ats_malloc tracks any size.

I thought malloc had no specific guarantees on alignment, but usually provides 8 byte alignment? Perhaps @bryancall can move the extra byes to the end?

Can you use `malloc_usable_size` on a specific category of memory allocation? Being able to pass in the counter here allows you to strictly control which allocations are being tracked.

**Comment by bgaff** [comment]
@PSUdaemon malloc guarantees 8 byte alignment, fwiw. This still perseveres 8 byte alignment assuming sizeof(size_t) remains 8 bytes.

**Comment by PSUdaemon** [comment]
@bgaff,

I think the 8 byte thing is a glibc implementation detail. Also, [this link](http://www.delorie.com/gnu/docs/glibc/libc_31.html
) suggests it's 16 bytes on 64bit as @jpeach suggested. Which yes, is still also 8 byte aligned. I think moving the size to the end of the allocated memory will give us the previous behavior.

**Comment by bgaff** [comment]
Ahh yes, you're correct @PSUdaemon . Nonetheless you cannot move the size_t to the end because how would you find it given a raw pointer? 

But @jpeach 's concern can be handled by allocating 16 bytes for tracking.

**Comment by PSUdaemon** [comment]
Duh. Sorry, looking too hard at malloc.

I have been trying to find macro we can use, but there isn't anything that I can find that works well. This is hidden inside `malloc.c` in glibc `#define MALLOC_ALIGNMENT       (2 * SIZE_SZ)`. Oh well, I guess just making it 16 bytes works. Feels hacky though.

**Comment by bgaff** [comment]
> Feels hacky though.

Maybe something like jemalloc would be better solution to this problem as @jpeach has suggested.

**Comment by dragon512** [comment]
I am not sure why we are using C++ and not using new to do this? We want to track memory in Traffic server that is being used to alloc memory for OpenSSL ( ie not in OpenSSL as we would have to do fancier instrumentation to do this right). We can just overload new so we code do a 

auto p = new(track::openssl) myobj;

This would be more functional and maintainable for this attempt.

**Comment by bryancall** [comment]
@jpeach 
I don't think it can get less ugly.  I tried to clean it up as much as possible.

What do you mean by duplicates the memory tracking that is already done?  There is no memory tracking for OpenSSL right now.  This is how I was able to see an issue with OpenSSL memory usage in our production environment and resolve the issue.

Is there an issue with memory alignment and is there a performance concern if we are using this as a diagnostic tool?

**Comment by dragon512** [comment]
I take back what I said after talking to Bryan call.. This is using a openssl lib callback API. So this looks good. The CPU is going to fine as long as we are on 32-bit bounds, so 8 or 16 bytes will be fine. There is just overhead with each alloc.

**Comment by bryancall** [comment]
@jpeach 
Do you know what the overhead is on malloc_usable_size().  Maybe switching over to that would be a better idea.

**Comment by jpeach** [comment]
> What do you mean by duplicates the memory tracking that is already done?

I just meant that the malloc implementation is already tracking the allocation size.

> Do you know what the overhead is on malloc_usable_size()?

The space overhead will be smaller, since malloc can know things that you don't and optimize away the need to carry the size with every allocation. For example, it can know that pointers within a certain range always point to 8 byte chunks.

I expect the time overhead on ``malloc_usable_size`` would be the same or better.

**Comment by bryancall** [comment]
Updated to use malloc_usable_size()

**Comment by bryancall** [comment]
@jpeach They are both counters.

@PSUdaemon Yes, old_size > new_size  :)

**Comment by bryancall** [comment]
@PSUdaemon said there is FreeBSD support for malloc_usable_size(), but not for OmniOS.

Need to #ifdef for OmniOS.

**Comment by bryancall** [comment]
Updated to look for malloc_usable_size in autoconf and #ifdef for support.

**Pull request closed without merging**